New: Preserve language tags when importing subtitle files

lyarenei opened the pull request:
#### Database Migration
YES

#### Description
I noticed too late in #2513 that while the import for multiple subtitle files for same language was fixed, the language tags (forced, sdh, default, ...) were still dropped. So these changes attempt to address that; preserve the tags when renaming/importing the subtitle files.

The recognized tags are in the [language regex](https://github.com/Sonarr/Sonarr/blob/0521fc5681e845030ac17a9ebf4664792b24e27f/src/NzbDrone.Core/Parser/LanguageParser.cs#L27).
I don't know if a support for more/custom tags is needed/wanted or not, for the sake of implementation simplicity, I assumed the language regex is enough.

#### Todos
- [x] Tests
- [ ] Wiki Updates


#### Issues Fixed or Closed by this PR
* https://github.com/Sonarr/Sonarr/issues/2570
    - Resolves the issue itself, but does not fix the language issue discussed later

* https://github.com/Sonarr/Sonarr/issues/3278 
    - This is already fixed by changes in #4768, but relates to the language tags topic.

lyarenei commented on src/NzbDrone.Core/Extras/Subtitles/SubtitleService.cs:
```diff
@@ -68,11 +68,18 @@ public override IEnumerable<ExtraFile> MoveFilesAfterRename(Series series, List<
 
             var movedFiles = new List<SubtitleFile>();
+            
+            string GetSubtitleFileIdentifier(SubtitleFile subtitleFile)
+            {
+                var languageTags = LanguageParser.ParseLanguageTags(subtitleFile.RelativePath);
+                return subtitleFile.Language + subtitleFile.Extension +
+                       languageTags.Aggregate(string.Empty, (tags, tag) => tags + tag);
+            }
```
Maybe it would be better to store the parsed tags in the database than repeatedly parsing them from the file name?

markus101 replied:
Pros and cons to both, if additional parsing is added later reparsing the filename will add more value, but not doing the same thing over and over is also valuable.

lyarenei replied:
I've left that as-is, since additional parsing is more likely to be added than not, imo. It will also leave the code "prepared" - minimizing possible future changes size, should someone implement something like that.

However, I'm still open to this and don't mind implementing this the other way.

lyarenei commented on src/NzbDrone.Core/Parser/LanguageParser.cs:
```diff
@@ -146,6 +148,23 @@ public static Language ParseSubtitleLanguage(string fileName)
+            try
+            {
+                var simpleFilename = Path.GetFileNameWithoutExtension(fileName);
+                var tagMatches = SubtitleLanguageRegex.Matches(simpleFilename);
+                var languageTags = from Match tagMatch in tagMatches select tagMatch.Groups["tags"].Value.ToLower();
```
I got some weird issue here as the code stopped returning multiple matches. The `.Matches()` just started behaving exactly like `.Match`, only giving the last match. Yesterday, it worked just fine.

I am not aware of any changes which could change the behavior, tried to clean and rebuild Sonarr, but it's still happening. Do you have any ideas?

lyarenei commented on src/NzbDrone.Core/Extras/Subtitles/SubtitleService.cs:
```diff
@@ -175,16 +183,18 @@ public override IEnumerable<ExtraFile> ImportFiles(LocalEpisode localEpisode, Ep
             }
 
-            var subtitleFiles = new List<Tuple<string, Language, string>>();
+            var subtitleFiles = new List<Tuple<string, Language, string, IEnumerable<string>>>();
```
I have been pondering about if the tags should be just put in the single string just after they are parsed (delimit them with . for example), but on the other hand having it as enumberable is more flexible.

Any opinion on that?

markus101 replied:
List/enumerable would be nicer to deal with, I'd also get away from the Tuple and move the return value to a class.

lyarenei replied:
Changed to class - f2df494f42333ff5d57baf3c5e5d620cf045c3d8
Tags are stored as enumerable and there's a shorthand for getting them as a string.

markus101 commented on src/NzbDrone.Core/Parser/LanguageParser.cs:
```diff
@@ -146,6 +148,23 @@ public static Language ParseSubtitleLanguage(string fileName)
+                Logger.Debug(ex, "Failed parsing language tags from subtitle file: {0}", fileName);
+            }
+
+            return new string[] {};
```
`IEnumerable.Empty` (not sure the exact syntax offhand)

lyarenei replied:
Fixed in 6f950a9.

markus101 commented on src/NzbDrone.Core/Parser/LanguageParser.cs:
```diff
@@ -146,6 +148,23 @@ public static Language ParseSubtitleLanguage(string fileName)
+                var simpleFilename = Path.GetFileNameWithoutExtension(fileName);
+                var match = SubtitleLanguageRegex.Match(simpleFilename);
+                var languageTags = from Capture tag in match.Groups["tags"].Captures select tag.Value.ToLower();
+                return languageTags.Where(tag => !tag.Empty());
```
Can empty tags be filtered before they're added to the enumerable?

lyarenei replied:
Yes - [b3ce4b9](https://github.com/Sonarr/Sonarr/pull/4924/commits/b3ce4b93fb91d0e3b5e8afac1c18b0f846257f64)

markus101 commented on src/NzbDrone.Core/Extras/Subtitles/SubtitleService.cs:
```diff
@@ -175,16 +183,18 @@ public override IEnumerable<ExtraFile> ImportFiles(LocalEpisode localEpisode, Ep
             }
 
-            var groupedSubtitleFiles = subtitleFiles.GroupBy(s => s.Item2 + s.Item3).ToList();
+            var groupedSubtitleFiles = subtitleFiles.GroupBy(s => s.Item2 + s.Item3 + s.Item4.Aggregate(string.Empty, (tags, tag) => tags + tag)).ToList();
```
This group-by is pretty rough as a Tuple, as mentioned above, convert to the return value to a class.

lyarenei replied:
Changed to class - https://github.com/Sonarr/Sonarr/commit/f2df494f42333ff5d57baf3c5e5d620cf045c3d8

markus101 commented on src/NzbDrone.Core/Extras/Subtitles/SubtitleService.cs:
```diff
@@ -235,6 +246,15 @@ private string GetSuffix(Language language, int copy, bool multipleCopies = fals
 
+            if (languageTags.Any())
+            {
+                foreach (var tag in languageTags)
```
I think you can shorten to:

```
suffixBuilder.Append(".");
suffixBuilder.Append(string.Join('.', languageTags));
```

lyarenei replied:
Indeed, refactored in [3b8ce68](https://github.com/Sonarr/Sonarr/pull/4924/commits/3b8ce68c56e17d8f09dda428c9bdc499e31e99a3)

markus101 reviewed (changes requested):
Haven't tested but added some comments and responded to your questions.

lyarenei commented:
Looking at the failing build - the integration tests for ubuntu are failing - more exactly there are 44 ignored test cases due to `ignore reason: OneTimeSetUp: httpbin is bugged`.

lyarenei commented:
@markus101 Thanks for the review and answers. I've addressed the comments, so you can take a look again.

Since this feature now contains a database migration, are ther any additional steps I need to do? I've looked into the contribution guidelines and didn't find anything, so I'm asking just in case.

lyarenei commented:
Hi @markus101, can you please take a look when you'll have some time?

markus101 reviewed (changes requested):
Storing the FullPath is a non-starter, if you need to pass the full path for import purposes passing it as a separate parameter or a property that isn't stored (named `Path` to match `EpisodeFile`) would be much better.

lyarenei commented on src/NzbDrone.Core/Datastore/Migration/170_add_fields_to_subtitle_file.cs:
```diff
@@ -0,0 +1,15 @@
+        protected override void MainDbUpgrade()
+        {
+            Alter.Table("SubtitleFiles").AddColumn("LanguageTags").AsString().Nullable();
+            Alter.Table("SubtitleFiles").AddColumn("FullPath").AsString().Nullable();
```
Refactored in 8c22325

lyarenei commented on src/NzbDrone.Core/Datastore/Migration/170_add_fields_to_subtitle_file.cs:
```diff
@@ -0,0 +1,15 @@
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(170)]
+    public class add_fields_to_subtitle_file : NzbDroneMigrationBase
```
Renamed in 0fad6dc

lyarenei commented on src/NzbDrone.Core/Parser/LanguageParser.cs:
```diff
@@ -152,6 +154,25 @@ public static Language ParseSubtitleLanguage(string fileName)
+            {
+                var simpleFilename = Path.GetFileNameWithoutExtension(fileName);
+                var match = SubtitleLanguageRegex.Match(simpleFilename);
+                var languageTags = from Capture tag in match.Groups["tags"].Captures
```
Rewritten in 57656c1

lyarenei commented on src/NzbDrone.Core/Extras/Subtitles/SubtitleService.cs:
```diff
@@ -235,6 +245,12 @@ private string GetSuffix(Language language, int copy, bool multipleCopies = fals
                 suffixBuilder.Append(IsoLanguages.Get(language).TwoLetterCode);
             }
 
+            if (!string.IsNullOrEmpty(languageTags))
```
Cool, didn't notice that, still new to C#. Thanks, fixed in cb36976.

lyarenei commented on src/NzbDrone.Core/Extras/Subtitles/SubtitleFile.cs:
```diff
@@ -1,10 +1,28 @@
+            LanguageTags = string.Join(".", newTags);
+        }
+
+        public void SetRelativePath(string fullPath, string seriesPath)
+        {
+            var endsWithSeparator = seriesPath.EndsWith(Path.PathSeparator.ToString());
+            var startPos = endsWithSeparator ? seriesPath.Length : seriesPath.Length + 1;
+            RelativePath = fullPath.Substring(startPos);
+        }
```
Was thinking about moving this to ExtraFile, but that might be better to move it there in the future if needed?

markus101 replied:
`PathExtensions.GetRelativePath()` already exists to handle this logic, `subFile.RelativePath = ...`

lyarenei commented:
> Storing the FullPath is a non-starter, if you need to pass the full path for import purposes passing it as a separate parameter or a property that isn't stored (named `Path` to match `EpisodeFile`) would be much better.

I've noticed `RelativePath` int the `ExtraFile` base class, so I ended up using that.

All comments should be resolved, please have a look again. Thanks.

markus101 commented on src/NzbDrone.Core/Extras/Subtitles/SubtitleFile.cs:
```diff
@@ -1,10 +1,28 @@
+
+        public string LanguageTags { get; set; }
+
+        public void SetLanguageTags(IEnumerable<string> newTags)
```
Language tags can be set directly, no need for a function to set it.

lyarenei replied:
Removed in 24334c4.

markus101 commented on src/NzbDrone.Core/Extras/Subtitles/SubtitleFile.cs:
```diff
@@ -1,10 +1,28 @@
+
+        public string AggregateString => Language + Extension + LanguageTags;
+
+        public string LanguageTags { get; set; }
```
Thinking about this a bit more, I think it'd make sense to make this an array and have the value stored in the DB as such (serialized as a JSON string), the logic to automatically serialize and deserialize already exists.

lyarenei replied:
Refactored in 24334c4.

markus101 commented on src/NzbDrone.Core/Extras/Subtitles/SubtitleFile.cs:
```diff
@@ -1,10 +1,28 @@
     {
         public Language Language { get; set; }
+
+        public string AggregateString => Language + Extension + LanguageTags;
```
Make it `Language + LanguageTags + Extension` to match naming?

lyarenei replied:
Reordered in 49d89e2.

markus101 reviewed (changes requested):
Thanks for the fixes, a few more things that will make things a bit more flexible in the future.

lyarenei commented on src/NzbDrone.Core/Datastore/Migration/add_language_tags_to_subtitle_files.cs:
```diff
@@ -0,0 +1,14 @@
+using FluentMigrator;
```
Oh I see, looks like I dropped it in 96a7c73e6a9b7934a5aa2c154ec06ea30e462f3c. Fixed in 139417e824521f287661c14f32c68acf0e3ed1b7.

lyarenei commented:
I'm not sure what to do with the failing tests. When I tried to take a look and run them locally, I got over 100 failures and all of them seem not even remotely related to these changes - for example I have ~70 failures in API tests. 

Any ideas what might be wrong?

lyarenei commented:
> Sorry, those tests were broken in develop and went unnoticed. If you rebase on develop they'll be resolved. As for the local tests I'm not sure why they'd be failing.

Thanks for the fix. There seems to be only one test failing, but that one passes locally, so maybe some instability at play here? Everything else seems to look good.

markus101 commented:
Thanks for all your fixes on this!

lyarenei commented:
> Thanks for all your fixes on this!

@markus101 No problem. What about the wiki update, do you want to mention this change there? Should I edit the page (settings/importing) now or wait until a new release? Or is this handled by someone else?

markus101 commented:
I think it's worth adding there with a note that it's in develop only at the moment.

lyarenei commented:
> I think it's worth adding there with a note that it's in develop only at the moment.

Ok, I updated the wiki page. https://wiki.servarr.com/en/sonarr/settings#importing.
